src/lib/client/jubSignalClient.ts

```diff
@@ -0,0 +1,296 @@
+          const userId = await hashPublicKeyToUUID(metadata.fromPublicKey);
+          const user = users[userId];
+          if (user) {
+            user.name = metadata.fromDisplayName;
+            user.encPk = metadata.fromPublicKey;
+            user.x = x;
+            user.tg = tg;
+            user.sigPk = pk;
+            user.msg = msg;
+            user.sig = sig;
+            user.inTs = metadata.timestamp.toISOString();
+
+            users[userId] = user;
+          } else {
+            users[userId] = {
+              name: metadata.fromDisplayName,
+              encPk: metadata.fromPublicKey,
+              x,
+              tg,
+              sigPk: pk,
+              msg,
+              sig,
+              inTs: metadata.timestamp.toISOString(),
+            };
+          }
+
+          const activity = {
+            type: JUB_SIGNAL_MESSAGE_TYPE.INBOUND_TAP,
+            name: metadata.fromDisplayName,
+            id: userId,
+            ts: metadata.timestamp.toISOString(),
+          };
+          activities.push(activity);
+        } catch (error) {
+          console.error(
+            "Invalid inbound tap message received from server: ",
+            message
+          );
+        } finally {
+          break;
+        }
+      case JUB_SIGNAL_MESSAGE_TYPE.LOCATION_TAP:
+        try {
+          if (metadata.fromPublicKey !== recipientPublicKey) {
+            throw new Error(
+              "Invalid message: location tap messages must be sent from self"
+            );
+          }
+
+          const { id, name, pk, msg, sig } =
+            await locationTapMessageSchema.validate(data);
+          const location = locationSignatures[id];
+          if (location) {
+            location.id = id;
+            location.name = name;
+            location.pk = pk;
+            location.msg = msg;
+            location.sig = sig;
+            location.ts = metadata.timestamp.toISOString();
+
+            locationSignatures[id] = location;
+          } else {
+            locationSignatures[id] = {
+              id,
+              name,
+              pk,
+              msg,
+              sig,
+              ts: metadata.timestamp.toISOString(),
+            };
+          }
+
+          const activity = {
+            type: JUB_SIGNAL_MESSAGE_TYPE.LOCATION_TAP,
+            name,
+            id,
+            ts: metadata.timestamp.toISOString(),
+          };
+          activities.push(activity);
+        } catch (error) {
+          console.error(
+            "Invalid location tap message received from server: ",
+            message
+          );
+        } finally {
+          break;
+        }
+      case JUB_SIGNAL_MESSAGE_TYPE.QUEST_COMPLETED:
+        console.error("Unable to handle quest completed messages");
+        break;
+      default:
+        console.error("Received invalid message type");
+    }
+  }
+
+  activities.reverse(); // We want activities to be in reverse chronological order
+
```

Comment:
We should make some sort of test suite to ensure this properly merges when multiple taps happen in a row. I'm not sure it will behave super well across tabs